Use smart pointers for ScriptExecutionContext and CanvasRenderingContext (m_context)

chirags27 opened the pull request:
#### e233ab4568b25c084d63488de1ea116928aadc13
<pre>
Use smart pointers for ScriptExecutionContext and CanvasRenderingContext (m_context)
<a href="https://bugs.webkit.org/show_bug.cgi?id=254347">https://bugs.webkit.org/show_bug.cgi?id=254347</a>
rdar://107142041

Reviewed by Chris Dumez and Ryosuke Niwa.

This change adopts Ref&apos;s for m_context member variable to align with the
smart pointer usage guidelines.

* Source/WTF/wtf/CheckedRef.h:
(WTF::downcast):
* Source/WTF/wtf/Ref.h:
(WTF::downcast):
* Source/WTF/wtf/RefPtr.h:
(WTF::downcast):
* Source/WTF/wtf/WeakPtr.h:
(WTF::downcast):
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::processPendingTasks):
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::shadowRealmLoader const):
(WebCore::ScriptModuleLoader::resolve):
(WebCore::ScriptModuleLoader::fetch):
(WebCore::ScriptModuleLoader::moduleURL):
(WebCore::ScriptModuleLoader::responseURLFromRequestURL):
(WebCore::ScriptModuleLoader::evaluate):
(WebCore::ScriptModuleLoader::importModule):
(WebCore::ScriptModuleLoader::notifyFinished):
* Source/WebCore/bindings/js/ScriptModuleLoader.h:
* Source/WebCore/dom/RejectedPromiseTracker.cpp:
(WebCore::RejectedPromiseTracker::promiseHandled):
(WebCore::RejectedPromiseTracker::processQueueSoon):
(WebCore::RejectedPromiseTracker::reportUnhandledRejections):
(WebCore::RejectedPromiseTracker::reportRejectionHandled):
* Source/WebCore/dom/RejectedPromiseTracker.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::InspectorCanvas):
(WebCore::InspectorCanvas::canvasElement const):
(WebCore::InspectorCanvas::scriptExecutionContext const):
(WebCore::InspectorCanvas::resolveContext const):
(WebCore:: const):
(WebCore::InspectorCanvas::canvasChanged):
(WebCore::InspectorCanvas::resetRecordingData):
(WebCore::InspectorCanvas::buildObjectForCanvas):
(WebCore::InspectorCanvas::buildInitialState):
* Source/WebCore/inspector/InspectorCanvas.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimerFireState::DOMTimerFireState):
(WebCore::DOMTimerFireState::~DOMTimerFireState):

Canonical link: <a href="https://commits.webkit.org/262252@main">https://commits.webkit.org/262252@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cba4ccb9a311393b5f4c3656b701a083a419b1bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/955 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1019 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1411 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/848 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1067 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1062 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/904 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/897 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1317 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/949 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/894 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/891 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/847 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/872 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/921 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1946 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/938 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/913 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/866 "4 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/983 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/860 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/903 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/197 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/249 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/927 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/986 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/191 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.h:
```diff
@@ -90,7 +90,7 @@ class ThreadableWebSocketChannelClientWrapper : public ThreadSafeRefCounted<Thre
     void processPendingTasks();
 
-    ScriptExecutionContext& m_context;
+    Ref<ScriptExecutionContext> m_context;
```
These references probably need to be WeakPtr, not RefPtr.

chirags27 replied:
I'm not sure if that is correct - it currently holds a raw reference, so it's assuming the object to be alive? (so the changes should probably make that assumption true?)

cdumez replied:
The previous code didn't look safe to me since ThreadableWebSocketChannelClientWrapper is ref counted. Someone could totally keep it alive longer than the Document.

Also note that switching a `T&` to a `Ref<T>&` is not always safe since it can introduce reference cycles and thus leaks.

chirags27 replied:
Ok, is the suggestion to change to WeakPtr for only this, or everywhere in this change? It's tricky to figure out the expected lifetime just by looking at the member variable, so I appreciate any inputs there.

rniwa replied:
Everywhere.

rniwa commented on Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
```diff
@@ -257,7 +257,7 @@ void ThreadableWebSocketChannelClientWrapper::processPendingTasks()
         // When a synchronous operation is in progress (i.e. the execution stack contains
         // WorkerThreadableWebSocketChannel::waitForMethodCompletion()), we cannot invoke callbacks in this run loop.
-        m_context.postTask([this, protectedThis = Ref { *this }] (ScriptExecutionContext& context) {
+        m_context->postTask([this, protectedThis = Ref { *this }] (ScriptExecutionContext& context) {
```
We should store m_context in RefPtr before calling this function. And ditto for other code in this patch.

rniwa commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -268,11 +268,11 @@ JSC::JSValue ScriptModuleLoader::evaluate(JSC::JSGlobalObject* jsGlobalObject, J
         RELEASE_AND_RETURN(scope, moduleRecord->evaluate(m_shadowRealmGlobal, awaitedValue, resumeMode));
     else if (m_ownerType == OwnerType::Document) {
-        if (auto* frame = downcast<Document>(m_context).frame())
+        if (auto* frame = downcast<Document>(m_context)->frame())
```
We should be using RefPtr for frame.

rniwa commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -82,7 +82,7 @@ ScriptModuleLoader::~ScriptModuleLoader()
 UniqueRef<ScriptModuleLoader> ScriptModuleLoader::shadowRealmLoader(JSC::JSGlobalObject* realmGlobal) const
 {
-    auto loader = makeUniqueRef<ScriptModuleLoader>(m_context, m_ownerType);
+    auto loader = makeUniqueRef<ScriptModuleLoader>(*m_context, m_ownerType);
```
We need to first store m_context in Ref/RefPtr before using it.
So do: Ref { *m_context }

rniwa commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -124,7 +124,7 @@ JSC::Identifier ScriptModuleLoader::resolve(JSC::JSGlobalObject* jsGlobalObject,
     RETURN_IF_EXCEPTION(scope, { });
 
-    auto result = resolveModuleSpecifier(m_context, m_ownerType, jsGlobalObject->importMap(), specifier, baseURL);
+    auto result = resolveModuleSpecifier(*m_context, m_ownerType, jsGlobalObject->importMap(), specifier, baseURL);
```
Same goes for all other places we're currently doing *m_context.

rniwa commented on Source/WebCore/page/DOMTimer.cpp:
```diff
@@ -93,7 +93,7 @@ class DOMTimerFireState {
 
 private:
-    ScriptExecutionContext& m_context;
+    WeakPtr<ScriptExecutionContext> m_context;
```
I guess this one can be RefPtr since DOMTimerFireState is always allocated in stack.

rniwa commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -200,16 +203,16 @@ JSC::JSInternalPromise* ScriptModuleLoader::fetch(JSC::JSGlobalObject* jsGlobalO
         auto loader = CachedModuleScriptLoader::create(*this, deferred.get(), *static_cast<CachedScriptFetcher*>(JSC::jsCast<JSC::JSScriptFetcher*>(scriptFetcher)->fetcher()), WTFMove(parameters));
         m_loaders.add(loader.copyRef());
-        if (!loader->load(downcast<Document>(m_context), WTFMove(completedURL))) {
+        if (!loader->load(*downcast<Document>(m_context), WTFMove(completedURL))) {
```
Why not use protectedContext here as well.

rniwa commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -487,16 +492,16 @@ void ScriptModuleLoader::notifyFinished(ModuleScriptLoader& moduleScriptLoader,
                 if (!matchIntegrityMetadata(cachedScript, integrity)) {
-                    m_context.addConsoleMessage(MessageSource::Security, MessageLevel::Error, makeString("Cannot load script ", integrityMismatchDescription(cachedScript, integrity)));
-                    rejectWithFetchError(m_context, WTFMove(promise), TypeError, "Cannot load script due to integrity mismatch"_s);
+                    m_context->addConsoleMessage(MessageSource::Security, MessageLevel::Error, makeString("Cannot load script ", integrityMismatchDescription(cachedScript, integrity)));
```
We should use protectedContext here as well.

rniwa commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -591,7 +596,7 @@ void ScriptModuleLoader::notifyFinished(ModuleScriptLoader& moduleScriptLoader,
     }
 
-    m_context.eventLoop().queueTask(TaskSource::Networking, [promise = WTFMove(promise), sourceCode = WTFMove(sourceCode)]() {
+    m_context->eventLoop().queueTask(TaskSource::Networking, [promise = WTFMove(promise), sourceCode = WTFMove(sourceCode)]() {
```
Use protectedContext?

ddkilzer commented on Source/WTF/wtf/WeakPtr.h:
```diff
@@ -338,6 +338,18 @@ inline bool is(const WeakPtr<ArgType, WeakPtrImpl>& source)
+inline Target* downcast(const WeakPtr<Source, WeakPtrImpl>& source)
+{
+    return static_cast<Target*>(source.get());
+}
```
We should never allow this `downcast()` specialization for a `ThreadSafeWeakPtr<>` type as that introduces a race condition between the time the raw pointer is returned, and before it's stored in a `RefPtr<>`.

Maybe this is already the case, though since `WeakPtr<>::get()` returns a `RefPtr<>` instead of a raw pointer, so [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae) prevents this from being used.

Also, the pattern for `WeakPtr<>` requires:

1. Getting the raw pointer from `WeakPtr<>` and storing in (non-temporary!) `RefPtr<>`.
2. Checking if `RefPtr<>` is not `nullptr`.
3. Getting raw pointer from `RefPtr<>` to use in a method call if item 2 is true.

This applies to all `WeakPtr<>`, not just `ThreadSafeWeakPtr<>`.

```
    WeakPtr<Foo> weakFoo;
    RefPtr refFoo = weakFoo.get();
    if (refFoo)
        callSomeMethodWithFooPointer(downcast(refFoo));
```

chirags27 replied:
WeakPtr::get currently returns a raw pointer AFAICT. I'm not sure if I understand why this downcast is invalid/ worse than what we already have today.

cdumez replied:
Correct, WeakPtr::get() returns a raw pointer. As a matter of fact, WeakPtr works with types that are not refcounted so you couldn't use David's pattern in all cases.

David's comment mentioned ThreadSafeWeakPtr, which is not impacted by your template specialization so maybe there is a misunderstanding.

cdumez replied:
For what it's worth, I think this downcast<> specialization for WeakPtr<> is fine.

I also agree with David that we shouldn't add such specialization for ThreadSafeWeakPtr but I don't think that's relevant to this patch.

ddkilzer commented on Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
```diff
@@ -257,16 +257,18 @@ void ThreadableWebSocketChannelClientWrapper::processPendingTasks()
         // WorkerThreadableWebSocketChannel::waitForMethodCompletion()), we cannot invoke callbacks in this run loop.
-        m_context.postTask([this, protectedThis = Ref { *this }] (ScriptExecutionContext& context) {
+        RefPtr protectedContext = m_context.get();
+        protectedContext->postTask([this, protectedThis = Ref { *this }] (ScriptExecutionContext& context) {
```
We need to check if `protectedContext` is `nullptr` here since it's coming from a `WeakPtr<>`.

ddkilzer commented on Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
```diff
@@ -257,16 +257,18 @@ void ThreadableWebSocketChannelClientWrapper::processPendingTasks()
 
     Vector<std::unique_ptr<ScriptExecutionContext::Task>> pendingTasks = WTFMove(m_pendingTasks);
+    Ref protectedContext = { *m_context };
     for (auto& task : pendingTasks)
-        task->performTask(m_context);
+        task->performTask(protectedContext.get());
```
Need to add `nullptr` check for `protectedContext` since it's coming from a `WeakPtr<>`, unless the method call handles `nullptr` values.

ddkilzer commented on Source/WebCore/bindings/js/ScriptModuleLoader.h:
```diff
@@ -59,7 +59,7 @@ class ScriptModuleLoader final : private ModuleScriptLoaderClient {
     UniqueRef<ScriptModuleLoader> shadowRealmLoader(JSC::JSGlobalObject* realmGlobal) const;
 
-    ScriptExecutionContext& context() { return m_context; }
+    ScriptExecutionContext& context() { return *m_context; }
```
The `context()` method should return `RefPtr<>` since `m_context` is `WeakPtr<>`, and could now be `nullptr`.

Dereferencing `nullptr` is undefined behavior here.

ddkilzer commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -454,22 +458,23 @@ void ScriptModuleLoader::notifyFinished(ModuleScriptLoader& moduleScriptLoader,
     };
 
     JSC::SourceCode sourceCode;
+    Ref protectedContext = { *m_context };
```
Need to add `nullptr` checks below since `protectedContext` could be `nullptr` now, unless each method can handle a `nullptr` value being passed to it, or we don't care if we dereference `nullptr` and crash (less likely).

Could be an argument for returning early if `m_context` is `nullptr` here, too.  I'm not familiar enough with this code to know if that's an option.

NOTE: I'm not going to keep adding "Need `nullptr` check(s)." comments everywhere as it's already getting noisy.

ddkilzer commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -124,7 +125,8 @@ JSC::Identifier ScriptModuleLoader::resolve(JSC::JSGlobalObject* jsGlobalObject,
     RETURN_IF_EXCEPTION(scope, { });
 
-    auto result = resolveModuleSpecifier(m_context, m_ownerType, jsGlobalObject->importMap(), specifier, baseURL);
+    Ref protectedContext = { *m_context };
```
Need `nullptr` check(s) (see below).

ddkilzer commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -82,7 +82,8 @@ ScriptModuleLoader::~ScriptModuleLoader()
 {
-    auto loader = makeUniqueRef<ScriptModuleLoader>(m_context, m_ownerType);
+    Ref protectedContext = { *m_context };
+    auto loader = makeUniqueRef<ScriptModuleLoader>(protectedContext.get(), m_ownerType);
```
May need `nullptr` check if this method doesn't handle `nullptr` values.

ddkilzer commented on Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.h:
```diff
@@ -90,7 +90,7 @@ class ThreadableWebSocketChannelClientWrapper : public ThreadSafeRefCounted<Thre
     void processPendingTasks();
 
-    ScriptExecutionContext& m_context;
+    WeakPtr<ScriptExecutionContext> m_context;
```
Not directly related to this patch, but question for @cdumez, @achristensen07 or @rniwa:

Why is `ScriptExecutionContext` only `CanMakeWeakPtr<>` (and not `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<>`) when we have methods like `ScriptExecutionContext::postCrossThreadTask()` that use a `ScriptExecutionContext&`?

In other words, how do we guarantee that `ScriptExecutionContext` stays alive on background threads for lambdas being run asynchronously?

cdumez replied:
> Why is ScriptExecutionContext only CanMakeWeakPtr<> (and not ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<>) 

Probably because ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr is not new and we haven't ported ScriptExecutionContext to it yet.

Also, `ScriptExecutionContext::postCrossThreadTask()` is barely used. Only IDB uses it in one place, using a raw pointer to a ScriptExecutionContext. It looks like the code is trying to be safe at least using a Lock both when calling postCrossThreadTask() and when the ScriptExecutionContext is getting destroyed.

ddkilzer commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -175,21 +177,22 @@ JSC::JSInternalPromise* ScriptModuleLoader::fetch(JSC::JSGlobalObject* jsGlobalO
     auto* jsPromise = JSC::JSInternalPromise::create(vm, globalObject.internalPromiseStructure());
     RELEASE_ASSERT(jsPromise);
     auto deferred = DeferredPromise::create(globalObject, *jsPromise);
+    Ref protectedContext = { *m_context };
```
Need `nullptr` check(s) (see below).

ddkilzer commented on Source/WebCore/dom/RejectedPromiseTracker.cpp:
```diff
@@ -148,7 +150,7 @@ void RejectedPromiseTracker::reportUnhandledRejections(Vector<UnhandledPromise>&
     // https://html.spec.whatwg.org/multipage/webappapis.html#unhandled-promise-rejections
 
-    VM& vm = m_context.vm();
+    VM& vm = m_context->vm();
```
Why doesn't this `m_context` have a `protectedContext` like the others?

ddkilzer commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -107,18 +107,18 @@ Ref<InspectorCanvas> InspectorCanvas::create(CanvasRenderingContext& context)
 HTMLCanvasElement* InspectorCanvas::canvasElement() const
 {
-    return dynamicDowncast<HTMLCanvasElement>(m_context.canvasBase());
+    return dynamicDowncast<HTMLCanvasElement>(m_context->canvasBase());
```
This needs a `RefPtr<> protectedContext` to keep it alive during the method call, and possibly a `nullptr` check to prevent a crash.

Same for most uses of `m_context` in this class.

ddkilzer commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -892,7 +892,8 @@ Ref<Protocol::Canvas::Canvas> InspectorCanvas::buildObjectForCanvas(bool capture
 
-    if (auto attributes = buildObjectForCanvasContextAttributes(m_context))
+    Ref protectedContext = { *m_context };
+    if (auto attributes = buildObjectForCanvasContextAttributes(protectedContext.get()))
```
May need `nullptr` check if the method doesn't handle that.

ddkilzer commented on Source/WebCore/inspector/InspectorCanvas.h:
```diff
@@ -52,7 +52,7 @@ class InspectorCanvas final : public RefCounted<InspectorCanvas> {
     const String& identifier() const { return m_identifier; }
 
-    CanvasRenderingContext& canvasContext() const { return m_context; }
+    CanvasRenderingContext& canvasContext() const { return *m_context; }
```
This needs to return a `RefPtr< CanvasRenderingContext>`, not a reference.

This causes undefined behavior if `m_context` is `nullptr`.

ddkilzer commented on Source/WebCore/page/DOMTimer.cpp:
```diff
@@ -93,7 +93,7 @@ class DOMTimerFireState {
 
 private:
-    ScriptExecutionContext& m_context;
+    Ref<ScriptExecutionContext> m_context;
```
The constructor for this class should take a `Ref<ScriptExecutionContext>&&` and then use `WTFMove()` up on Line 59:

```
    DOMTimerFireState(Ref<ScriptExecutionContext>&& context, int nestingLevel)
        : m_context(WTFMove(context))
```

This will require changing call sites, but this way it's clear that `DOMTimer` is going to hold a non-`nullptr` reference to a `ScriptExecutionContext`.

ddkilzer commented on Source/WebCore/inspector/InspectorCanvas.h:
```diff
@@ -130,7 +130,7 @@ class InspectorCanvas final : public RefCounted<InspectorCanvas> {
     String m_identifier;
 
-    CanvasRenderingContext& m_context;
+    WeakPtr<CanvasRenderingContext> m_context;
```
Need `RefPtr<> protectedContext` and possibly `nullptr` checks where `m_context` is used in the class.

cdumez replied:
It bugs me that we're using a WeakPtr but keep assuming that it is non-null. Why didn't you use a CheckedRef<> here?

ddkilzer reviewed (changes requested):
Overall concerns:
- Getters need to return `RefPtr<Foo>` instead of `Foo&` to avoid undefined behavior.
- All uses of `m_context` in classes need a corresponding `protectedContext` and likely `nullptr` checks (whether that's an early return or checking at call sites).
- `DOMTimer` constructor argument should be changed.

Do _not_ need to fix with this patch, but an open question for other reviewers:
- Why isn't `ScriptExecutionContext` using `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<>`?

ddkilzer reviewed (commented):
Okay, if @rniwa was fine with the patch, I'm not going to second-guess here (so feel free to ignore my other comments), but I'd really like to understand what the rules are here.

1. When to add a `protectedFoo` before calling `WeakPtr<>::get()` to obtain a raw pointer?

I thought the point of this patch was to always add a `protectedFoo` in those cases since the methods still take raw pointers, and there are no guarantees that any given method won't cause the object to be freed that the raw pointer points to.

This is not done consistently in this patch currently.

2. When to add a `nullptr` check when after getting a raw pointer from a `WeakPtr<>`.

The prevailing direction here seems to be "only if it crashes later", or if there is some specific knowledge that the `WeakPtr<>` can be `nullptr`.  This ignores clear situations where undefined behavior may occur.  For example, dereferencing a pointer unconditionally lets the compiler assume the variable never contains a `nullptr` value, resulting in UB.  (But I guess that's no worse than it was before when the instance variables were references.)

cdumez commented on Source/WebCore/bindings/js/ScriptModuleLoader.h:
```diff
@@ -73,7 +73,7 @@ class ScriptModuleLoader final : private ModuleScriptLoaderClient {
     URL responseURLFromRequestURL(JSC::JSGlobalObject&, JSC::JSValue);
 
-    ScriptExecutionContext& m_context;
+    WeakPtr<ScriptExecutionContext> m_context;
```
ScriptModuleLoader is owned by the ScriptExecutionContext. The code was safe and a WeakPtr is not needed because m_context can never be null. @rniwa introduced CheckedRef<> specifically for such purposes. CheckedRef is more efficient here and makes it clear m_context cannot be null and it is safe to dereference. 

Same comment may apply in other places.

cdumez commented on Source/WebCore/dom/RejectedPromiseTracker.h:
```diff
@@ -59,7 +59,7 @@ class RejectedPromiseTracker {
     void reportRejectionHandled(Ref<DOMPromise>&&);
 
-    ScriptExecutionContext& m_context;
+    WeakPtr<ScriptExecutionContext> m_context;
```
This is also a clear use case for `CheckedRef<ScriptExecutionContext>` since m_context is the owner and can never be null (cc @rniwa who wrote CheckedRef).

cdumez replied:
Same comment may apply in other places in this patch.

cdumez commented on Source/WTF/wtf/CheckedRef.h:
```diff
@@ -196,6 +196,18 @@ inline bool is(const CheckedRef<ArgType, ArgPtrTraits>& source)
+}
+
+template<typename ExpectedType, typename ArgType, typename ArgPtrTraits>
+inline ExpectedType& downcast(const CheckedRef<ArgType, ArgPtrTraits>& source)
```
Seems this should return a `const ExpectedType&` no? Since CheckedRef::get() returns a `const T&` when called on a const CheckedRef.

cdumez commented on Source/WTF/wtf/Ref.h:
```diff
@@ -276,6 +276,18 @@ inline bool is(const Ref<ArgType, PtrTraits>& source)
+}
+
+template<typename Target, typename Source, typename PtrTraits>
+inline Target& downcast(const Ref<Source, PtrTraits>& source)
```
Here `Target&` instead of `const Target&` is fine because Ref::get() returns a `T&` when called on a const Ref.

cdumez commented on Source/WTF/wtf/Ref.h:
```diff
@@ -276,6 +276,18 @@ inline bool is(const Ref<ArgType, PtrTraits>& source)
+template<typename Target, typename Source, typename PtrTraits>
+inline Target& downcast(Ref<Source, PtrTraits>& source)
+{
+    return static_cast<Target&>(source.get());
```
You should probably be calling downcast<>() here. Otherwise, this downcast() version does not assertion checking, which defeats the purpose of downcast.

cdumez commented on Source/WTF/wtf/Ref.h:
```diff
@@ -276,6 +276,18 @@ inline bool is(const Ref<ArgType, PtrTraits>& source)
+template<typename Target, typename Source, typename PtrTraits>
+inline Target& downcast(const Ref<Source, PtrTraits>& source)
+{
+    return static_cast<Target&>(source.get());
```
should call downcast.

cdumez commented on Source/WTF/wtf/RefPtr.h:
```diff
@@ -282,6 +282,18 @@ inline bool is(const RefPtr<ArgType, PtrTraits, RefDerefTraits>& source)
+template<typename Target, typename Source, typename PtrTraits, typename RefDerefTraits>
+inline Target* downcast(RefPtr<Source, PtrTraits, RefDerefTraits>& source)
+{
+    return static_cast<Target*>(source.get());
```
Should call downcast.

cdumez commented on Source/WTF/wtf/RefPtr.h:
```diff
@@ -282,6 +282,18 @@ inline bool is(const RefPtr<ArgType, PtrTraits, RefDerefTraits>& source)
+template<typename Target, typename Source, typename PtrTraits, typename RefDerefTraits>
+inline Target* downcast(const RefPtr<Source, PtrTraits, RefDerefTraits>& source)
+{
+    return static_cast<Target*>(source.get());
```
should call downcast.

cdumez commented on Source/WTF/wtf/WeakPtr.h:
```diff
@@ -338,6 +338,18 @@ inline bool is(const WeakPtr<ArgType, WeakPtrImpl>& source)
+template<typename Target, typename Source, typename WeakPtrImpl>
+inline Target* downcast(WeakPtr<Source, WeakPtrImpl>& source)
+{
+    return static_cast<Target*>(source.get());
```
Should call downcast.

cdumez commented on Source/WTF/wtf/WeakPtr.h:
```diff
@@ -338,6 +338,18 @@ inline bool is(const WeakPtr<ArgType, WeakPtrImpl>& source)
+template<typename Target, typename Source, typename WeakPtrImpl>
+inline Target* downcast(const WeakPtr<Source, WeakPtrImpl>& source)
+{
+    return static_cast<Target*>(source.get());
```
Should call downcast.

cdumez commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -107,18 +107,18 @@ Ref<InspectorCanvas> InspectorCanvas::create(CanvasRenderingContext& context)
 InspectorCanvas::InspectorCanvas(CanvasRenderingContext& context)
     : m_identifier("canvas:" + IdentifiersFactory::createIdentifier())
-    , m_context(context)
+    , m_context(&context)
```
This change is not needed and may cause an unnecessary null check.

cdumez commented on Source/WebCore/page/DOMTimer.cpp:
```diff
@@ -93,7 +93,7 @@ class DOMTimerFireState {
 
 private:
-    ScriptExecutionContext& m_context;
+    Ref<ScriptExecutionContext> m_context;
```
DOMFireState is a stack object so this seems safe.

chirags27 replied:
That's correct, but nothing stops this from being created on the heap. I think the intent is to convert member variables that can support smart pointers to adopting them.

cdumez commented on Source/WTF/wtf/Ref.h:
```diff
@@ -276,6 +276,18 @@ inline bool is(const Ref<ArgType, PtrTraits>& source)
+}
+
+template<typename Target, typename Source, typename PtrTraits>
+inline const Target& downcast(const Ref<Source, PtrTraits>& source)
```
Why does this return a const?

cdumez commented on Source/WTF/wtf/RefPtr.h:
```diff
@@ -282,6 +282,18 @@ inline bool is(const RefPtr<ArgType, PtrTraits, RefDerefTraits>& source)
+}
+
+template<typename Target, typename Source, typename PtrTraits, typename RefDerefTraits>
+inline const Target* downcast(const RefPtr<Source, PtrTraits, RefDerefTraits>& source)
```
Why is this returning a const?

cdumez commented on Source/WTF/wtf/WeakPtr.h:
```diff
@@ -338,6 +338,18 @@ inline bool is(const WeakPtr<ArgType, WeakPtrImpl>& source)
+}
+
+template<typename Target, typename Source, typename WeakPtrImpl>
+inline const Target* downcast(const WeakPtr<Source, WeakPtrImpl>& source)
```
Why does this return a const?

cdumez commented on Source/WebCore/html/canvas/CanvasRenderingContext.h:
```diff
@@ -49,7 +49,7 @@ class SVGImageElement;
 enum class PixelFormat : uint8_t;
 
-class CanvasRenderingContext : public ScriptWrappable {
+class CanvasRenderingContext : public ScriptWrappable, public CanMakeWeakPtr<CanvasRenderingContext>, public CanMakeCheckedPtr {
```
Should include <wtf/CheckedRef.h> since this is where CanMakeCheckedPtr is defined.

cdumez commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -113,40 +113,40 @@ InspectorCanvas::InspectorCanvas(CanvasRenderingContext& context)
 {
     JSC::JSLockHolder lock(exec);
-
+    auto& context = const_cast<CanvasRenderingContext&>(m_context.get());
```
Have you tried updating the InspectorCanvas::resolveContext() function to not be `const` instead of doing a const_cast here?

chirags27 replied:
You're right, that would work, I'll fix this

cdumez reviewed (approved):
LGTM with nits.

chirags27 committed:
Use smart pointers for ScriptExecutionContext and CanvasRenderingContext (m_context)
https://bugs.webkit.org/show_bug.cgi?id=254347
rdar://107142041

Reviewed by Chris Dumez and Ryosuke Niwa.

This change adopts Ref's for m_context member variable to align with the
smart pointer usage guidelines.

* Source/WTF/wtf/CheckedRef.h:
(WTF::downcast):
* Source/WTF/wtf/Ref.h:
(WTF::downcast):
* Source/WTF/wtf/RefPtr.h:
(WTF::downcast):
* Source/WTF/wtf/WeakPtr.h:
(WTF::downcast):
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::processPendingTasks):
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::shadowRealmLoader const):
(WebCore::ScriptModuleLoader::resolve):
(WebCore::ScriptModuleLoader::fetch):
(WebCore::ScriptModuleLoader::moduleURL):
(WebCore::ScriptModuleLoader::responseURLFromRequestURL):
(WebCore::ScriptModuleLoader::evaluate):
(WebCore::ScriptModuleLoader::importModule):
(WebCore::ScriptModuleLoader::notifyFinished):
* Source/WebCore/bindings/js/ScriptModuleLoader.h:
* Source/WebCore/dom/RejectedPromiseTracker.cpp:
(WebCore::RejectedPromiseTracker::promiseHandled):
(WebCore::RejectedPromiseTracker::processQueueSoon):
(WebCore::RejectedPromiseTracker::reportUnhandledRejections):
(WebCore::RejectedPromiseTracker::reportRejectionHandled):
* Source/WebCore/dom/RejectedPromiseTracker.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::InspectorCanvas):
(WebCore::InspectorCanvas::canvasElement const):
(WebCore::InspectorCanvas::scriptExecutionContext const):
(WebCore::InspectorCanvas::resolveContext const):
(WebCore:: const):
(WebCore::InspectorCanvas::canvasChanged):
(WebCore::InspectorCanvas::resetRecordingData):
(WebCore::InspectorCanvas::buildObjectForCanvas):
(WebCore::InspectorCanvas::buildInitialState):
* Source/WebCore/inspector/InspectorCanvas.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimerFireState::DOMTimerFireState):
(WebCore::DOMTimerFireState::~DOMTimerFireState):

Canonical link: https://commits.webkit.org/262252@main
